### PR TITLE
feat: Add support for all listener actions (redirect, fixed-response, authenticate-cognito, authenticate-oidc)

### PR DESCRIPTION
**Pull request by rriveramdz**
# PR o'clock

## Description

Extend current LB to support all listener actions with all its properties.

Current behavior is the same. When no `action_type` specified, will always default to forward action so no breaking changes.

### Checklist

* [x] `terraform fmt` and `terraform validate` both work from the root and `examples/alb_test_fixture` directories (look in CI for an example)
* [x] Tests for the changes have been added and passing (for bug fixes/features)
* [ ] Test results are pasted in this PR (in lieu of CI)
* [x] Docs have been added/updated (for bug fixes/features)
* [x] Any breaking changes are noted in the description above

### CODE REVIEW

**Comment by rriveramdz** [comment]
Hello Anton!

Thanks for the suggestion, I wasn't sure there could be nested dynamic blocks. I just updated the draft. I think it should be very improved now (not sure if totally correct). 

I'll start working in extending the documentation and adding examples for creating each kind of listener.

**Comment by rriveramdz** [comment]
Hey Anton!

I've run into an issue regarding a dynamic block in the authenticated listeners:

```hcl
dynamic "authentication_request_extra_params" {
            for_each = ......
            content {
              authentication_request_extra_params.key = authentication_request_extra_params.value
            }
}
```

I was trying to do something like that, and also tried some other options but haven't found yet a way to pass the map set into that block. Have you ever run into this scenario?

**Comment by antonbabenko** [comment]
I am not sure what exactly you try to do but I am quiet sure you will find all combinations required for `dynamic` blocks in the s3-bucket module.

Also, if you push examples it is easier to understand what you want to do.

**Comment by rriveramdz** [comment]
I just added some examples. The problem I'm facing is that there's a dynamic block inside the authenticated listener actions, but there's no defined set of keys, these are user provided and I haven't been able to figure out a way to dynamically grab each key-value pair and render it into the block.

**Comment by antonbabenko** [comment]
I still don't understand. Could you please write some code for me and the desired output, so that maybe I understand what you have to generate faster?

I don't have a lot of time to read docs, debug, write code for this part myself, sorry.

**Comment by rriveramdz** [comment]
This is a `authenticate_oidc` block inside the listener:

```hcl
dynamic "authenticate_oidc" {
        for_each = length(keys(lookup(default_action.value, "authenticate_oidc_block", {}))) == 0 ? [] : [lookup(default_action.value, "authenticate_oidc_block", {})]

        # Max 10 extra params
        content {
          dynamic "authentication_request_extra_params" {
            for_each = length(keys(lookup(authenticate_oidc.value, "authentication_request_extra_params", {}))) == 0 ? [] : [lookup(authenticate_oidc.value, "authentication_request_extra_params", {})]

            content {
              authentication_request_extra_params.key = authentication_request_extra_params.value
          }
        }
          authorization_endpoint     = lookup(authenticate_oidc.value, "authorization_endpoint", null)
          client_id                  = lookup(authenticate_oidc.value, "client_id", null)
          client_secret              = lookup(authenticate_oidc.value, "client_secret", null)
          issuer                     = lookup(authenticate_oidc.value, "issuer", null)
          on_unauthenticated_request = lookup(authenticate_oidc.value, "on_unauthenticated_request", null)
          scope                      = lookup(authenticate_oidc.value, "scope", null)
          session_cookie_name        = lookup(authenticate_oidc.value, "session_cookie_name", null)
          session_timeout            = lookup(authenticate_oidc.value, "session_timeout", null)
          token_endpoint             = lookup(authenticate_oidc.value, "token_endpoint", null)
          user_info_endpoint         = lookup(authenticate_oidc.value, "user_info_endpoint", null)
        }
      }
```

My problem is that the property `authentication_request_extra_params` is a map of user defined properties and I want to render all keys and values provided in this configuration:

```json
{
      port            = 83
      protocol     = "HTTP"
      action_type  = "authenticate-cognito"
      target_group_index = 0
      authenticate_cognito_block = {
        authentication_request_extra_params = {
          display = "page"
          prompt  = "login"
          property = "value"
          property2 = "value2"
          property3 = "value3"
        }
        on_unauthenticated_request = "authenticate"
        session_cookie_name        = "session-${random_pet.this.id}"
        session_timeout            = 3600
        user_pool_arn              = aws_cognito_user_pool.user_pool.arn
        user_pool_client_id        = aws_cognito_user_pool_client.user_pool_client.id
        user_pool_domain           = aws_cognito_user_pool_domain.user_pool_domain.domain
      }
    }
```

The result that I'd like to have is to render the user provided map:

authentication_request_extra_params {
          display = "page"
          prompt  = "login"
          property = "value"
          property2 = "value2"
          property3 = "value3"
}

But I haven't been able to set the keys provided from the user, dynamic expects that the key is already known. I hope this clarifies a little more. Thanks for the help.

**Comment by antonbabenko** [comment]
Thanks for the explanation! It makes a lot of sense to me now.

I think you are right, there is no way to render content of a block with configurable keys but...

As I can see in the test - https://github.com/terraform-providers/terraform-provider-aws/blob/866c3d335d7a9b6a2d3aa25f71a03324a380266b/aws/resource_aws_lb_listener_test.go#L1198-L1200 

```
authentication_request_extra_params = {
    param = "test"
}
```

This means that `authentication_request_extra_params` is not a block, but is a `map`. Which means you should be able to render map dynamically.

**Comment by rriveramdz** [comment]
Okay, then I think it's done. I'm just about to test it and write the documentation and examples. Thanks!

**Comment by rriveramdz** [comment]
Hey Anton!

I'm having another issue regarding the attribute types, as you remember the way I defined the other listener actions follows this structure:

```hcl
{
      port            = 83
      protocol     = "HTTP"
      action_type  = "authenticate-cognito"
      target_group_index = 0
      authenticate_cognito_block = {
        authentication_request_extra_params = {
          display = "page"
          prompt  = "login"
          property = "value"
          property2 = "value2"
          property3 = "value3"
        }
        on_unauthenticated_request = "authenticate"
        session_cookie_name        = "session-${random_pet.this.id}"
        session_timeout            = 3600
        user_pool_arn              = aws_cognito_user_pool.user_pool.arn
        user_pool_client_id        = aws_cognito_user_pool_client.user_pool_client.id
        user_pool_domain           = aws_cognito_user_pool_domain.user_pool_domain.domain
      }
   }
```

According to the original code this is a `list(map(string))`. However, there can be at most two nested maps inside the structure. I tried changing the type to `list(map(any))` but it always throws an error that _map items must be of the same type_. I checked some of the other modules and found that in those cases you used the `any` type, but it's hard to iterate over it as Terraform creates dynamically a tuple with all of the child types necessary. I've found hard to keep the code as is now without modifying that much. Do you happen to have any suggestions? Thanks.

**Comment by antonbabenko** [comment]
You are right, I use `type = any` in such cases.

If you still have time after this PR is done, you can look into using `try()` to solve type conversion issues but I don't think this should be done in the module (at least not before Terraform 0.13 is out).

File: main.tf [hunk 1]
```diff
@@ -130,9 +166,81 @@ resource "aws_lb_listener" "frontend_https" {
...
+        }
+      }
+
+      dynamic "authenticate_oidc" {
+        for_each = length(keys(lookup(default_action.value, "authenticate_oidc", {}))) == 0 ? [] : [lookup(default_action.value, "authenticate_oidc", {})]
+
+        content {
+          # Max 10 extra params
+          authentication_request_extra_params = lookup(authenticate_oidc.value, "authentication_request_extra_params", null)
+          authorization_endpoint              = lookup(authenticate_oidc.value, "authorization_endpoint", null)
```
Comment (antonbabenko) [inline main.tf:223]: Some of the arguments are required (for eg, `authorization_endpoint`), so we can't use `null` for them.

It is better to write `authorization_endpoint = authenticate_oidc.value["authorization_endpoint"]`, and let it fail when user don't specify it during plan.

Please review and update all arguments of actions blocks as specified here: https://www.terraform.io/docs/providers/aws/r/lb_listener.html

Comment (rriveramdz) [reply main.tf:223]: Haven't really though about that. Just updated it.

**Comment by antonbabenko** [comment]
[v5.3.0](https://github.com/terraform-aws-modules/terraform-aws-alb/releases/v5.3.0) has been just released!

Thanks a lot, Raul!

**Pull request merged**